Introduce markdownlinting

khiga8 opened the pull request:
Related: https://github.com/github/accessibility/issues/2542

This PR enables markdown linting pulling in configs from [markdownlint-github](https://github.com/github/markdownlint-github). 

_Why?_ This repo is primarily composed of markdown files. There are valuable markdownlint rules that encourage best accessibility practices. We have config recs centralized in `markdownlint-github` now that should be pulled in.

`markdownlint-github` configs enables the [`markdownlint` defaults](https://github.com/DavidAnson/markdownlint/blob/main/schema/.markdownlint.jsonc), but then adds some GitHub custom rules and overrides. Some of markdownlint's defaults are related to stylistic things like `line-length` and `blanks-around-headings`. I don't know how much value Primer gets in enforcing these stylistic things so I decided to turn them off. There's also a lot to resolve, but if you think they should be enabled, it should probably be done in a follow-up.

I made sure the rules pertaining to accessibility are kept on since these provide a lot of value and encourage accessibility practices. I think those should be addressed as part of this PR. Open to reviewer opinions!

khiga8 commented on .markdownlint-cli2.cjs:
```diff
@@ -0,0 +1,19 @@
+
+const options = githubMarkdownOpinions.init({
+  // Disable rules we don't currently care to enforce pertaining to stylistic things.
+  'line-length': false,
+  'blanks-around-headings': false,
+  'blanks-around-lists': false,
+  'no-trailing-spaces': false,
+  'no-multiple-blanks': false,
+  'no-trailing-punctuation': false,
+  'single-trailing-newline': false,
+  'ul-indent': false,
+  'no-hard-tabs': false
```
From PR description: 

Some of markdownlint's defaults are related to stylistic things like `line-length` and `blanks-around-headings`. I don't know how much value Primer gets in enforcing these things so I decided to turn them off. There's also a lot to resolve.

I made sure the rules pertaining to accessibility are kept on since these provide a lot of value in encouraging accessibility practices.

khiga8 commented:
Upon running `markdownlint` based on these configs, I get these failures:

```
content/components/component-documentation-guidelines/component-documentation-guidelines.md:22 MD001/heading-increment/header-increment Heading levels should only increment by one level at a time [Expected: h3; Actual: h4]
content/components/data-table.mdx:167:8 MD042/no-empty-links No empty links [Context: "[Blankslate]()"]
content/foundations/color.mdx:35 MD001/heading-increment/header-increment Heading levels should only increment by one level at a time [Expected: h3; Actual: h4]
content/foundations/color.mdx:77 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:78 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:79 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:80 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:84 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:85 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:86 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:87 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:88 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:89 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:90 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:91 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:99 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:100 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:101 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:102 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:106 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:107 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:108 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:109 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:110 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:111 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:112 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:113 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:114 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:122 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:123 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:127 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:128 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:129 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:130 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:131 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:132 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:133 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:134 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/color.mdx:135 MD045/no-alt-text Images should have alternate text (alt text)
content/foundations/typography.mdx:32 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]
content/foundations/typography.mdx:38 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]
content/foundations/typography.mdx:44 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]
content/ui-patterns/button-usage.mdx:119 MD001/heading-increment/header-increment Heading levels should only increment by one level at a time [Expected: h3; Actual: h4]
content/ui-patterns/empty-states.mdx:32 GH002/no-generic-link-text Avoid using generic link text like `Learn more` or `Click here` [For link: Learn more]
content/ui-patterns/empty-states.mdx:32:215 MD042/no-empty-links No empty links [Context: "[Learn more about X](#)"]
content/ui-patterns/empty-states.mdx:32:244 MD042/no-empty-links No empty links [Context: "[Check out the guide on X](#)"]
content/ui-patterns/empty-states.mdx:32:286 MD042/no-empty-links No empty links [Context: "[Learn more](#)"]
content/ui-patterns/feature-onboarding.mdx:136 MD024/no-duplicate-heading/no-duplicate-header Multiple headings with the same content [Context: "### Usage guidelines"]
content/ui-patterns/feature-onboarding.mdx:316 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading [Context: "In-product marketing empty sta..."]
content/ui-patterns/forms.mdx:38:41 MD051/link-fragments Link fragments should be valid [Context: "[label](#label)"]
content/ui-patterns/forms.mdx:107:130 MD051/link-fragments Link fragments should be valid [Context: "[autocomplete](#autocomplete)"]
content/ui-patterns/messaging.mdx:63 GH002/no-generic-link-text Avoid using generic link text like `Learn more` or `Click here` [For link: here]
docs/content.md:5:1 MD019/no-multiple-space-atx Multiple spaces after hash on atx style heading [Context: "###  Principles"]
```

khiga8 commented on .github/workflows/deploy.yml:
```diff
@@ -17,7 +17,7 @@ jobs:
     uses: primer/.github/.github/workflows/deploy.yml@0cec9b9914f358846163f2428663b58da41028c9
     with:
-      node_version: 12
+      node_version: 14
```
I upgraded the `node_version` for compatability with `markdownlint-cli2` node dependency. I saw `ci.yml` already uses `node_version: 14` so assuming this is fine...

khiga8 commented:
I'd like a first-pass review of this PR from Primer reviewers before fixing!

jonrohan commented on package.json:
```diff
@@ -7,7 +7,7 @@
     "build:preview": "gatsby build",
     "now-build": "yarn build",
-    "lint": "eslint ."
+    "lint": "eslint . & markdownlint-cli2 \"**/*.{md,mdx}\" \"!.github\" \"!node_modules\""
```
Not a blocker, but I like when the `ci.yml` workflow is split up. currently it says linting "all" but it would be more helpful if we had jobs for "eslint" and "markdownlint" so the error is more clear.

jonrohan reviewed (approved):
Nice

khiga8 commented:
Addressed @jonrohan feedback and split into two jobs.

I addressed all the violations. You can go through each one starting from https://github.com/primer/design/pull/345/commits/cb4ccc1cd45af942d6a81b5aca919108a2d52f5c.

khiga8 commented on content/ui-patterns/feature-onboarding.mdx:
```diff
@@ -133,7 +133,7 @@ Treat feature onboarding as a story or a journey with a beginning, middle, and e
 A teaching bubble is a [popover](https://primer.style/css/components/popover) that calls attention to a feature in a specific part of the page. Generally, teaching bubbles should be used to educate the user and enrich the task at hand.
 
-### Usage guidelines
+### Usage guidelines for teaching bubble
```
This one was to address: [no-duplicate-heading](https://github.com/updownpress/markdown-lint/blob/master/rules/024-no-duplicate-header.md).

The rationale is:

> Some markdown parses generate anchors for headers based on the header name, and having headers with the same content can cause problems with this.

which was new to me! Sure enough, we have this issue with [usage guidelines](https://primer.style/design/ui-patterns/feature-onboarding#usage-guidelines.
